paystack/src/main/java/co/paystack/android/TransactionManager.java

```diff
@@ -105,180 +87,171 @@ void chargeCard(Activity activity, Charge charge, Paystack.TransactionCallback t
-        this.transactionCallback.showLoading(true);
-        HashMap<String, String> params = validateRequestBody.getParamsHashMap();
-        Call<TransactionApiResponse> call = apiService.validateCharge(params);
-        call.enqueue(serverCallback);
-    }
+        if (chargeResponse.getAuth() != null && !chargeResponse.getAuth().equalsIgnoreCase("none")) {
+            authenticateTransaction(chargeParams, chargeResponse, transaction);
+            return;
+        }
 
-    private void reQueryChargeOnServer() throws KeyManagementException, NoSuchAlgorithmException, KeyStoreException {
-        this.transactionCallback.showLoading(true);
-        Call<TransactionApiResponse> call = apiService.requeryTransaction(transaction.getId());
-        call.enqueue(serverCallback);
+        setProcessingOff();
+        notifyProcessingError(new ChargeException(chargeResponse.getMessage()));
     }
 
-    private void initiateChargeOnServer() throws KeyManagementException, NoSuchAlgorithmException, KeyStoreException {
-        this.transactionCallback.showLoading(true);
-        Call<TransactionApiResponse> call = apiService.charge(chargeRequestBody.getParamsHashMap());
-        call.enqueue(serverCallback);
+    private void authenticateTransaction(ChargeParams chargeParams, ChargeResponse chargeResponse, Transaction transaction) {
+        String authType = chargeResponse.getAuth();
+        assert authType != null;
 
-    }
-
-    private void handleApiResponse(TransactionApiResponse transactionApiResponse) {
-        if (transactionApiResponse == null) {
-            transactionApiResponse = TransactionApiResponse.unknownServerResponse();
+        String authMessage = "Authentication required";
+        if (chargeResponse.getOtpMessage() != null) {
+            authMessage = chargeResponse.getOtpMessage();
+        } else if (chargeResponse.getMessage() != null) {
+            authMessage = chargeResponse.getMessage();
         }
 
-        // The AVS charge endpoint sends an "errors" object when address verification fails
-        if (transactionApiResponse.hasErrors) {
-            notifyProcessingError(new ChargeException(transactionApiResponse.message));
-            return;
-        }
-        transaction.loadFromResponse(transactionApiResponse);
-
-        if (transactionApiResponse.status.equalsIgnoreCase("1") || transactionApiResponse.status.equalsIgnoreCase("success")) {
+        if (authType.equalsIgnoreCase(AuthType.ADDRESS_VERIFICATION)) {
+            new AddressVerificationAsyncTask(chargeParams).execute(chargeResponse.getCountryCode());
+        } else if (authType.equalsIgnoreCase(AuthType.PIN)) {
+            new PinAsyncTask(chargeParams).execute();
+        } else if (authType.equalsIgnoreCase(AuthType.OTP) || authType.equalsIgnoreCase(AuthType.PHONE)) {
+            transactionCallback.beforeValidate(transaction);
+            osi.setOtpMessage(authMessage);
```

Comment:
Relatively minor, but we only use `authMessage` for `PHONE`,`OTP` and `3DS`, so it feels like we could preserve the immutability of `authMessage` by simply doing a null check when we do `asi.setUrl()` and in the event of null, we provide the fallback message.

Reply:
@nathanstasin I'm not sure I understand this

Reply:
In kotlin it would be: `osi.setOtpMessage(chargeResponse.getOtpMessage() ?: FALLBACK_AUTH_REQUIRED_MESSAGE)`

`authMessage` is a var when it could be a val, to use kotlin terminology. But, since this will hopefully be the last version before rewriting, I guess we can let it slide.